### PR TITLE
Enhance telemetry with server information

### PR DESCRIPTION
**Pull request by bryophyta**
## What does this change?

Add some fields to our client side telemetry:

- Pass a request id as a header to the server when making `fetch` requests, which as of #871 should be picked up and used by the server's own logging.
- Include that request id in our `NEWSWIRES_FETCHED_RESULTS` telemetry logging, allowing us to correlate the telemetry from the client with logging from the server.
- Pass a `Server-Timing` header back from the client which includes the server-side request duration, using [the standardised format](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Server-Timing) that developer tools know how to interpret.
- Extract the server timing from the response, and include it in our telemetry, so that now we have `duration` (total round trip from making the request to processing the response, in the client app) and `serverTiming` (the time spent processing the request on the server).

## How to test

- [x] Added some unit tests to the `extractServerTiming` function, to make sure that it handles edge cases without throwing.
- [x] Deployed to CODE and verified that the new logging shows up (see screenshot below).
- [x] I've also tried making requests with the network being throttled on the client, so that there's a much longer total duration but a normal server time. We can see that there's a long `duration` and a short `serverTiming` in the logs, which suggests that it's tracking what it's supposed to be.

## Images

image

image

## Introducing DB migrations?

Database migrations need to be applied manually before releasing. The docs can be found in db/README.md 
- [ ] Database migrations have been applied for CODE
- [ ] Database migrations have been applied for PROD

### CODE REVIEW

File: newswires/client/src/context/fetchResults.ts [hunk 1]
```diff
@@ -10,14 +12,21 @@ export const fetchResults = async ({
...
 	abortController?: AbortController;
-}): Promise<WiresQueryData> => {
+	requestId?: string;
+}): Promise<{
+	data: WiresQueryData;
+	requestId: string;
+	serverTiming?: number;
+}> => {
 	const endpoint = view.includes('dotcopy') ? '/api/dotcopy' : '/api/search';
+	const requestIdToUse = requestId ?? uuidv4();
```
Comment (lindseydew) [inline newswires/client/src/context/fetchResults.ts]: I found it a bit confusing that that this request id is passed as a param with a dynamic default if not. 

Is the idea that a new uuid is generated when it's called in application logic, and passed in explicitly for testing? 

Might be better to generate the uuidv4() on invocation and mock for a test? 

But let me know if I am missing something

File: newswires/app/lib/RequestLoggingFilter.scala [hunk 2]
```diff
@@ -39,17 +39,17 @@ class RequestLoggingFilter(implicit
 
     val resultFuture = next(withID)
 
-    resultFuture onComplete {
+    resultFuture.transform {
       case Success(response) =>
         val duration = System.currentTimeMillis() - start
         log(withID, Right(response), duration)
+        Success(response.withHeaders("Server-Timing" -> s"total;dur=$duration"))
```
Comment (lindseydew) [inline newswires/app/lib/RequestLoggingFilter.scala:46]: Curious as to why the server is just sending the total time when parsing unit tests include a database time as well? Is this in preparation for a possible extension?

File: newswires/client/src/context/fetchResults.ts [hunk 3]
```diff
@@ -46,8 +56,40 @@ export const fetchResults = async ({
...
+		requestId: requestIdToUse,
+		serverTiming,
 	};
 };
+
+function extractServerTiming(headers: Headers | undefined): number | undefined {
+	const serverTimingHeader = headers?.get('Server-Timing');
+	const serverTimingMatch = serverTimingHeader?.match(
+		/total;dur=(\d+(\.\d+)?)/,
+	);
```
Comment (lindseydew) [inline newswires/client/src/context/fetchResults.ts]: Think this method may be a bit more extensible if it parsed the Headers into a key value pair instead. For example:

```

function parseServerTiming(header) {
	if (!header) return [];

	return header.split(',').map(entry => {
		const [name, ...params] = entry.trim().split(';');

		const result = { name };

		for (const param of params) {
			const [key, value] = param.split('=');

			result[key] = value?.replace(/^"|"$/g, '');
		}

		return result;
	});
}
```

If this intention is to include the db times at some point. 

If not though, might be simpler for the server to just return the time and simplify the header structure?

File: newswires/client/src/context/fetchResults.ts [hunk 4]
```diff
@@ -46,8 +56,40 @@ export const fetchResults = async ({
...
+		return serverTiming;
+	} catch (e) {
+		console.error(
+			`Error extracting Server-Timing from response: ${getErrorMessage(e)}`,
+		);
+		return undefined;
+	}
+}
+
+export const forTesting = {
```
Comment (lindseydew) [inline newswires/client/src/context/fetchResults.ts]: I would just export the method directly so that we don't have to worry about a testing branch in the application code

File: newswires/client/src/context/SearchContext.tsx [hunk 5]
```diff
@@ -611,10 +619,12 @@ function createFetchedResultsTelemetryData({
 			]),
 		),
 		duration: performance.now() - startedAt,
+		...(serverTiming !== undefined ? { serverTiming } : {}),
```
Comment (lindseydew) [inline newswires/client/src/context/SearchContext.tsx:622]: Slightly simpler: 
```
...(serverTiming !== undefined && { serverTiming })
```

File: newswires/client/src/context/fetchResults/fetchResults.ts [hunk 6]
```diff
@@ -46,8 +56,40 @@ export const fetchResults = async ({
...
+
+	const serverTiming = extractServerTiming(response.headers);
+
 	return {
-		...parseResult.data,
-		results: parseResult.data.results.map(transformWireItemQueryResult),
+		data: {
+			...parseResult.data,
+			results: parseResult.data.results.map(transformWireItemQueryResult),
+		},
```
Comment (lindseydew) [inline newswires/client/src/context/fetchResults/fetchResults.ts:63]: Think it would make sense to group these into request metadata: 
meta: {
		requestId,
		serverTiming,
	},

File: newswires/client/src/context/SearchContext.tsx [hunk 7]
```diff
@@ -276,14 +276,16 @@ export function SearchContextProvider({ children }: PropsWithChildren) {
 			const startedAt = performance.now();
 			refreshesSinceLastTelemetrySend.current = 0;
 			fetchResults({ query: currentConfig.query, view: currentConfig.view })
-				.then((data) => {
+				.then(({ data, requestId, serverTiming }) => {
 					sendTelemetryEvent(
 						'NEWSWIRES_FETCHED_RESULTS',
 						createFetchedResultsTelemetryData({
```
Comment (lindseydew) [inline newswires/client/src/context/SearchContext.tsx:282]: Keep thinking this method could be worth a unit test at some point, but not necessarily in scope for this change

**Review by lindseydew — COMMENTED** [review]
Nice looks like a really good initiative. Just a few minor coding style suggestions

File: newswires/client/src/context/fetchResults.ts (hunk 3 above)
Comment (bryophyta) [reply newswires/client/src/context/fetchResults.ts]: I was only planning to pass the total duration for now, but the reason for the specific format here is to match the 'timing-metric' format used by browsers: https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Server-Timing#timing-metric

This allows the metrics to show up nicely in DevTools (screenshot in the PR description)

File: newswires/app/lib/RequestLoggingFilter.scala (hunk 2 above)
Comment (bryophyta) [reply newswires/app/lib/RequestLoggingFilter.scala:46]: The `db` metric in the unit tests is just there as an example of another possible metric, so that we can verify the function is picking out the `total` specifically, rather than other metrics which might be present in the header using the same format.

File: newswires/client/src/context/fetchResults.ts (hunk 1 above)
Comment (bryophyta) [reply newswires/client/src/context/fetchResults.ts]: > Is the idea that a new uuid is generated when it's called in application logic, and passed in explicitly for testing?

yep that was my reasoning 👍

I don't love mocking `uuidv4` for the tests, because I don't think we want to be too tied to it as an implementation detail, and also it seems conceivable that other internals might use `uuidv4` at some point, in which case they would also be affected. 

So I've opted to wrap the call in a separate `generateRequestId` function which we can mock instead. What do you think?

73c1f8c0

File: newswires/client/src/context/SearchContext.tsx (hunk 5 above)
Comment (bryophyta) [reply newswires/client/src/context/SearchContext.tsx:622]: agree that it's simpler, but I think I'd prefer not using `&&` here, because technically this ends up evaluating to `...(false)` if `serverTiming === undefined`, which means that when we read it we need to remember that `{ ...false }` is equivalent to `{ ...{} }`, but I think the latter is more intuitive

File: newswires/app/lib/RequestLoggingFilter.scala (hunk 2 above)
Comment (lindseydew) [reply newswires/app/lib/RequestLoggingFilter.scala:46]: Cool, make sense. My only very minor nit to follow this would be for the first test to verify the behaviour would be to have a unit test for what the server will actually send, and perhaps renaming the second metric `db` to something like `test` or `dummy` metric, just because I got the impression from the tests that the intended server response would be 
`'server-timing': 'total;dur=123.45, db;dur=67.89'`

File: newswires/client/src/context/fetchResults/generateRequestId.ts [hunk 8]
```diff
@@ -0,0 +1,8 @@
+import { v4 as uuidv4 } from 'uuid';
+
+/** Just a simple wrapper around uuidv4 to make it easier to mock in tests.
+ * Avoids the test setup needing to know about the internal implementation of fetchResults.
+ */
+export function generateRequestId(): string {
```
Comment (lindseydew) [inline newswires/client/src/context/fetchResults/generateRequestId.ts:6]: Nice, yeah I think this wrapper makes sense 👍

File: newswires/client/src/context/fetchResults/extractServerTiming.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
+import { getErrorMessage } from '@guardian/libs';
+
+export function extractServerTiming(
+	headers: Headers | undefined,
+): number | undefined {
+	const serverTimingHeader = headers?.get('Server-Timing');
+	const serverTimingMatch = serverTimingHeader?.match(
+		/total;dur=(\d+(\.\d+)?)/,
+	);
+	try {
```
Comment (lindseydew) [inline newswires/client/src/context/fetchResults/extractServerTiming.ts:10]: Small nit, 

```
  if(!serverTimingMatch) return undefined
  const serverTiming  = Number(serverTimingMatch[1])
  if (Number.isNaN(serverTiming)) return undefined
  return serverTiming
```

Might be slightly easier to parse

**Review by lindseydew — APPROVED** [review]
Looks great, great initiative 💯
